am-lib/src/main/java/uk/gov/hmcts/reform/amlib/internal/FilterService.java

```diff
@@ -20,38 +21,82 @@
+        List<JsonPointer> nodesToRemove = filterPointersWithoutReadPermission(attributePermissions);
+        log.debug("> Nodes without READ access: " + nodesToRemove);
+
+        nodesToKeep = filterNodesBySecurityClassification(nodesToKeep, nodesToRemove,
+            attributeSecurityClassifications, userSecurityClassifications);
+
+        if (!nodesToKeep.contains(WHOLE_RESOURCE_POINTER)) {
+            List<JsonPointer> uniqueNodesWithRead = reducePointersToUniqueList(nodesToKeep);
             log.debug("> Unique nodes with READ access: " + uniqueNodesWithRead);
 
             retainFieldsWithReadPermission(resourceCopy, uniqueNodesWithRead);
         }
 
-        List<JsonPointer> nodesWithoutRead = filterPointersWithoutReadPermission(attributePermissions);
-        log.debug("> Nodes without READ access: " + nodesWithoutRead);
+        if (nodesToKeep.isEmpty()) {
+            return null;
+        }
 
-        removeFieldsWithoutReadPermission(resourceCopy, nodesWithRead, nodesWithoutRead);
+        removeFieldsFromData(resourceCopy, nodesToKeep, nodesToRemove);
 
         return resourceCopy;
     }
 
+    private List<JsonPointer> filterNodesBySecurityClassification(List<JsonPointer> nodesToPotentiallyKeep,
+                                                     List<JsonPointer> nodesToRemove,
+                                                     Map<JsonPointer, SecurityClassification>
+                                                         attributeSecurityClassifications,
+                                                     Set<SecurityClassification> userSecurityClassifications) {
+        List<JsonPointer> nodesToKeep = new ArrayList<>();
+        nodesToPotentiallyKeep.forEach(node -> {
+            SecurityClassification nodeSecurityClassification = attributeSecurityClassifications.get(node);
+            if (nodeSecurityClassification == null) {
+                nodeSecurityClassification = inheritAttributeSecurityClassification(
+                    node, attributeSecurityClassifications);
+            }
+
+            if (userSecurityClassifications.contains(nodeSecurityClassification)) {
+                nodesToKeep.add(node);
+            } else {
+                log.debug("> Node without security classification access: " + node);
+                nodesToRemove.add(node);
+            }
+        });
+        return nodesToKeep;
+    }
+
+    private SecurityClassification inheritAttributeSecurityClassification(JsonPointer attribute,
+                                                                          Map<JsonPointer, SecurityClassification>
+                                                                              attributeSecurityClassifications) {
+        JsonPointer parentAttribute = attribute.head();
+        while (attributeSecurityClassifications.get(parentAttribute) == null) {
+            if (parentAttribute.toString().equals("")) {
```

Comment:
Same as above avoid checks for hard code. Null and empty string

Reply:
👍